File: backend/service/referenceCategoryService/referenceCategoryService.go
package referenceCategoryService

import (
	"errors"
	"permit-app/model"
	"permit-app/repo/moduleRepository"
	"permit-app/repo/referenceCategoryRepository"

	"gorm.io/gorm"
)

type ReferenceCategoryService interface {
	CreateReferenceCategory(req model.ReferenceCategoryRequest) (*model.ReferenceCategoryResponse, error)
	GetReferenceCategories(req model.ReferenceCategoryListRequest) ([]model.ReferenceCategoryResponse, int64, error)
	GetReferenceCategoryByID(id int64) (*model.ReferenceCategoryResponse, error)
	GetCategoriesByModuleID(moduleID int64) ([]model.ReferenceCategoryResponse, error)
	UpdateReferenceCategory(id int64, req model.ReferenceCategoryRequest) (*model.ReferenceCategoryResponse, error)
	DeleteReferenceCategory(id int64) error
}

type referenceCategoryServiceImpl struct {
	categoryRepo referenceCategoryRepository.ReferenceCategoryRepository
	moduleRepo   moduleRepository.ModuleRepository
}

func NewReferenceCategoryService(
	categoryRepo referenceCategoryRepository.ReferenceCategoryRepository,
	moduleRepo moduleRepository.ModuleRepository,
) ReferenceCategoryService {
	return &referenceCategoryServiceImpl{
		categoryRepo: categoryRepo,
		moduleRepo:   moduleRepo,
	}
}

func (s *referenceCategoryServiceImpl) CreateReferenceCategory(req model.ReferenceCategoryRequest) (*model.ReferenceCategoryResponse, error) {
	// Validate module exists
	_, err := s.moduleRepo.FindByID(req.ModuleID)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, errors.New("module not found")
		}
		return nil, err
	}

	category := &model.ReferenceCategory{
		ModuleID: req.ModuleID,
		Name:     req.Name,
		IsActive: true,
	}

	if req.IsActive != nil {
		category.IsActive = *req.IsActive
	}

	if err := s.categoryRepo.Create(category); err != nil {
		return nil, err
	}

	// Reload with relations
	category, err = s.categoryRepo.FindByID(category.ID)
	if err != nil {
		return nil, err
	}

	response := model.ToReferenceCategoryResponse(category)
	return &response, nil
}

func (s *referenceCategoryServiceImpl) GetReferenceCategories(req model.ReferenceCategoryListRequest) ([]model.ReferenceCategoryResponse, int64, error) {
	filters := make(map[string]interface{})

	if req.ModuleID > 0 {
		filters["module_id"] = req.ModuleID
	}
	if req.Name != "" {
		filters["name"] = req.Name
	}
	if req.IsActive != nil {
		filters["is_active"] = *req.IsActive
	}

	categories, total, err := s.categoryRepo.FindAll(filters, req.Page, req.Limit)
	if err != nil {
		return nil, 0, err
	}

	// Convert to response
	responses := make([]model.ReferenceCategoryResponse, len(categories))
	for i, category := range categories {
		responses[i] = model.ToReferenceCategoryResponse(&category)
	}

	return responses, total, nil
}

func (s *referenceCategoryServiceImpl) GetReferenceCategoryByID(id int64) (*model.ReferenceCategoryResponse, error) {
	category, err := s.categoryRepo.FindByID(id)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, errors.New("reference category not found")
		}
		return nil, err
	}

	response := model.ToReferenceCategoryResponse(category)
	return &response, nil
}

func (s *referenceCategoryServiceImpl) GetCategoriesByModuleID(moduleID int64) ([]model.ReferenceCategoryResponse, error) {
	categories, err := s.categoryRepo.FindByModuleID(moduleID)
	if err != nil {
		return nil, err
	}

	responses := make([]model.ReferenceCategoryResponse, len(categories))
	for i, category := range categories {
		responses[i] = model.ToReferenceCategoryResponse(&category)
	}

	return responses, nil
}

func (s *referenceCategoryServiceImpl) UpdateReferenceCategory(id int64, req model.ReferenceCategoryRequest) (*model.ReferenceCategoryResponse, error) {
	category, err := s.categoryRepo.FindByID(id)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, errors.New("reference category not found")
		}
		return nil, err
	}

	// Validate module exists if changed
	if req.ModuleID != category.ModuleID {
		_, err := s.moduleRepo.FindByID(req.ModuleID)
		if err != nil {
			if errors.Is(err, gorm.ErrRecordNotFound) {
				return nil, errors.New("module not found")
			}
			return nil, err
		}
	}

	category.ModuleID = req.ModuleID
	category.Name = req.Name
	if req.IsActive != nil {
		category.IsActive = *req.IsActive
	}

	if err := s.categoryRepo.Update(category); err != nil {
		return nil, err
	}

	// Reload with relations
	category, err = s.categoryRepo.FindByID(category.ID)
	if err != nil {
		return nil, err
	}

	response := model.ToReferenceCategoryResponse(category)
	return &response, nil
}

func (s *referenceCategoryServiceImpl) DeleteReferenceCategory(id int64) error {
	category, err := s.categoryRepo.FindByID(id)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return errors.New("reference category not found")
		}
		return err
	}

	return s.categoryRepo.Delete(category.ID)
}

File: frontend/src/services/referenceCategory.service.ts
import { apiClient } from "@/lib/api-client";
import { ApiResponse } from "@/types/api";
import {
  ReferenceCategory,
  ReferenceCategoryRequest,
  ReferenceCategoryListRequest,
} from "@/types/reference";

export const referenceCategoryService = {
  async getAll(
    params: ReferenceCategoryListRequest = {}
  ): Promise<
    ApiResponse<ReferenceCategory[]> & {
      meta?: { page: number; limit: number; total: number };
    }
  > {
    const queryParams = new URLSearchParams();

    if (params.module_id)
      queryParams.append("module_id", params.module_id.toString());
    if (params.name) queryParams.append("name", params.name);
    if (params.is_active !== undefined)
      queryParams.append("is_active", params.is_active.toString());
    if (params.page) queryParams.append("page", params.page.toString());
    if (params.limit) queryParams.append("limit", params.limit.toString());

    const response = await apiClient.get<ApiResponse<ReferenceCategory[]>>(
      `/reference-categories?${queryParams.toString()}`
    );
    return response as any;
  },

  async getById(id: number): Promise<ReferenceCategory> {
    const response = await apiClient.get<ApiResponse<ReferenceCategory>>(
      `/reference-categories/${id}`
    );
    return response.data;
  },

  async getByModuleId(moduleId: number): Promise<ReferenceCategory[]> {
    const response = await apiClient.get<ApiResponse<ReferenceCategory[]>>(
      `/modules/${moduleId}/categories`
    );
    return response.data;
  },

  async create(
    data: ReferenceCategoryRequest
  ): Promise<ReferenceCategory> {
    const response = await apiClient.post<ApiResponse<ReferenceCategory>>(
      "/reference-categories",
      data
    );
    return response.data;
  },

  async update(
    id: number,
    data: ReferenceCategoryRequest
  ): Promise<ReferenceCategory> {
    const response = await apiClient.put<ApiResponse<ReferenceCategory>>(
      `/reference-categories/${id}`,
      data
    );
    return response.data;
  },

  async delete(id: number): Promise<void> {
    await apiClient.delete(`/reference-categories/${id}`);
  },
};
